Source/Tools/FEXLoader/LinuxSyscalls/SignalDelegator.cpp

```diff
@@ -1376,13 +1383,105 @@ namespace FEX::HLE {
+          SigInfo.si_code == SEGV_ACCERR &&
+          SigInfo.si_addr == reinterpret_cast<void*>(Thread->CurrentFrame->State.DeferredSignalFaultAddress)) {
+        if (RefCount == 0) {
+          // We are faulting with an attempt to check deferred signals.
+          // Pull a signal frame off the stack.
+
+          mprotect(reinterpret_cast<void*>(Thread->CurrentFrame->State.DeferredSignalFaultAddress), 4096, PROT_READ | PROT_WRITE);
+
+          if (Thread->DeferredSignalFrames.size() == 0) {
+            // No signals to defer. Just set the fault page back to RW and continue execution.
+            // This occurs as a minor race condition between the refcount decrement and the access to the fault page.
+            return;
+          }
+          else {
+            auto Top = Thread->DeferredSignalFrames.back();
+            Signal = Top.Signal;
+            SigInfo = Top.Info;
+            Thread->DeferredSignalFrames.pop_back();
+
+            // Now that we are handling a deferred signal state. mprotect the fault page with RW permissions.
+            // This puts FEX in a state in-which we are *permanently* deferring signals and /not/ checking them.
+            //
+            // In order to return /back/ to a sane state, we wait for the rt_sigreturn to happen.
+            // rt_sigreturn will check if there are any more deferred signals to handle
+            // - If there are deferred signals
+            //   - mprotect back to PROT_NONE
+            //   - sigreturn will trampoline out to the previous fault address check, SIGSEGV and restart
+            // - If there are *no* deferred signals
+            //  - No need to mprotect, it is already RW
+          }
+        }
+        else {
+#ifdef _M_ARM_64
+          // If RefCount != 0 then that means we hit an access with stacked deferred signals.
+          // Increment the PC past the `str zr, [x1]` to continue code execution.
+          ArchHelpers::Context::SetPc(UContext, ArchHelpers::Context::GetPc(UContext) + 4);
+          return;
+#else
+          // X86 should always be doing a refcount compare and branch since we can't guarantee instruction size.
+          // ARM64 just always does the access to reduce branching overhead.
+          ERROR_AND_DIE_FMT("X86 shouldn't hit this DeferredSignalFaultAddress");
+#endif
+        }
+      }
+      else {
+        if (IsAsyncSignal(&SigInfo, Signal) && RefCount) {
+          // If the signal is asynchronous signal (as determined by si_code) and FEX is in a state of needing
+          // to defer the signal, then make sure to add the signal to the thread's signal queue.
+          Thread->DeferredSignalFrames.emplace_back(FEXCore::Core::InternalThreadState::DeferredSignalState {
+            .Info = SigInfo,
+            .Signal = Signal,
+          });
+
+          // Now update the faulting page permissions so it will fault on write.
+          // Just in case of multiple deferred signals, keep track of protection state.
```

Comment:
The kernel's tracking here is the authority because our tracking will end up getting decoupled at some point in the face of async signals occuring during our own internal signal handling. Recursive signals are a nightmare and depending on when it occurs we can't trust any state. Which is why all of this is such a faff to track to begin with.

Reply:
Ah, makes sense. Still a bit confused about the "Just in case of multiple deferred signals, keep track of protection state" bit though, since we aren't keeping track of anything but just telling the kernel to do things.

Reply:
Just deleted that line.